flink-runtime/src/main/java/org/apache/flink/runtime/management/JMXService.java

```diff
@@ -85,6 +86,9 @@ private static JMXServer startJMXServerWithPortRanges(Iterator<Integer> ports) {
 		while (ports.hasNext() && successfullyStartedServer == null) {
 			JMXServer server = new JMXServer();
 			int port = ports.next();
+			if (port == 0) { // try poke with a random port when port is set to zero
```

Comment:
How about we map 0 to a port-range (like 10000-65000) in `JMXService#startInstance`, and process it as usual? May not be as quick, but it should work more reliably.

As is we have to add in retries such that, if the port we deemed free was just taken, we try another port.

Reply:
actually this mapping to a large port range solution might be much more reliable, just one concern:

Since JMXServer part is on the container startup code path, I was wondering if we should set a timeout limit on the while loop - I haven't done any profiling but even if we assume each port poking takes ~1ms, this is still significant.

Reply:
That's a fair concern, and it does make me wonder whether we shouldn't just add a retry to the port poking approach; on average we'll likely end up with fewer attempts.

Nevertheless, I still thought about how we could alleviate the concern:
We could use a custom iterator that is not simply iterating over the port range, but also jumping into other port blocks.

So instead of checking `1, 2, 3, ..., 65000` we do ` 1000, 2000, 3000, ..., 65000, 1, 1001, 2001, ...`; i would think it is more likely that we run into a completely used port block (e.g., 8000-9000) than the first N ports in each block being used.
We can make the blocks arbitrarily small (e.g., 100), and the iterator implementation should be fairly simple.

```
non-iterator solution:

port = START // this could even be random value between 1 and BLOCK_SIZE
while (port != SOME_END_CRITERIA_PORT):
  if (!canCreateJMXServer(port)):
    port = port + BLOCK_SIZE
    if port > MAX:
      port = (port + 1) % MAX
```

Reply:
yeah. I think it is clear the solution is to map `port 0` to `port 10000 - 65000`. 
there are 2 potential todos here if we going down this path:
1. have a more randomized iterator just like you mentioned above, it could be put under `Netutils` that might be reused later.
2. with the randomized port iterator, do you think there's still need a hard timeout?

Reply:
I would think we wouldn't need the timeout.

Reply:
I would be hesitant in introducing this kind of magic with `0` mapping to `10000 - 65000`. Isn't it good enough to simply specify an explicit range if one runs multiple JMX services on the same node?

Reply:
> it would be better to do this conversion in the realm of the JMXServer
Agreed.

> Do we know that the startup period will be super long if we don't do this?
It probably isn't long. In the port-probing version, you are essentially entering a race condition between all processes on this node. I would consider it unlikely you will have more than a hundred of TaskExecutors on a given node, and only at that point should it be relevant in the first place.
I only suggested the special iterator to workaround the worst case scenario, something like the first 1000 ports already being allocated, where the current approach would likely fair miserably.
But there clear downsides to having this be applied in general to all users of the NetUtils; it is _very_ convenient to define a port-range and be pretty much guaranteed that all processes start at the front, and form a continuous port sequence if present on a single node.

Reply:
actually with some consideration, I dont think this PR is even necessary in a realistic stand point - setting the port to `"0"` vs setting it to some port range is pretty much equal in a platform job management standpoint. 

WRT the random port assignment, yeah I am not sure that's actually needed - we didn't do a profiling. 
I intent to close this PR and comment on the JIRA ticket if we all agree. (thank you so much for chiming in and sharing your thoughts @tillrohrmann @zentol  :-) )

Reply:
@walterddr Yes, I think we can close the PR. It is admittedly slightly less convenient to explicitly define a port range, but the users will also get more predictable behavior, and we don't have to deal with the added complexity.